app/src/behaviors/behavior_sticky_key.c

```diff
@@ -0,0 +1,274 @@
+    .binding_released = on_sticky_key_binding_released,
+};
+
+static int sticky_key_keycode_state_changed_listener(const struct zmk_event_header *eh) {
+    if (!is_keycode_state_changed(eh)) {
+        return 0;
+    }
+    struct keycode_state_changed *ev = cast_keycode_state_changed(eh);
+    for (int i = 0; i < ZMK_BHV_STICKY_KEY_MAX_HELD; i++) {
+        struct active_sticky_key *sticky_key = &active_sticky_key[i];
+        if (sticky_key->position == ZMK_BHV_STICKY_KEY_POSITION_NOT_USED) {
+            continue;
+        }
+
+        if (strcmp(sticky_key->config->behavior.behavior_dev, "KEY_PRESS") == 0 &&
+            HID_USAGE_ID(sticky_key->param1) == ev->keycode &&
+            HID_USAGE_PAGE(sticky_key->param1) == ev->usage_page &&
+            SELECT_MODS(sticky_key->param1) == ev->implicit_modifiers) {
+            // don't catch key down events generated by the sticky key behavior itself
+            continue;
+        }
+
+        // If events were queued, the timer event may be queued late or not at all.
+        // Release the sticky key if the timer should've run out in the meantime.
+        if (sticky_key->release_at != 0 && ev->timestamp > sticky_key->release_at) {
+            release_sticky_key_behavior(sticky_key, sticky_key->release_at);
+            if (stop_timer(sticky_key)) {
+                clear_sticky_key(sticky_key);
+            }
+            continue;
+        }
+
+        if (ev->state) { // key down
+            if (sticky_key->modified_key_usage_page != 0 || sticky_key->modified_key_keycode != 0) {
+                // this sticky key is already in use for a keycode
+                continue;
+            }
+            if (sticky_key->timer_started) {
+                stop_timer(sticky_key);
+            }
+            sticky_key->modified_key_usage_page = ev->usage_page;
+            sticky_key->modified_key_keycode = ev->keycode;
+
+        } else { // key up
+            if (sticky_key->timer_started &&
+                sticky_key->modified_key_usage_page == ev->usage_page &&
+                sticky_key->modified_key_keycode == ev->keycode) {
+                release_sticky_key_behavior(sticky_key, ev->timestamp);
+            }
+        }
+    }
+    return 0;
+}
+
+ZMK_LISTENER(behavior_sticky_key, sticky_key_keycode_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_sticky_key, keycode_state_changed);
+
+void behavior_sticky_key_timer_handler(struct k_work *item) {
+    struct active_sticky_key *sticky_key =
+        CONTAINER_OF(item, struct active_sticky_key, release_timer);
+    if (sticky_key->position == ZMK_BHV_STICKY_KEY_POSITION_NOT_USED) {
+        return;
+    }
+    if (!sticky_key->timer_is_cancelled) {
+        release_sticky_key_behavior(sticky_key, k_uptime_get());
+    }
+    clear_sticky_key(sticky_key);
+}
+
+static int behavior_sticky_key_init(struct device *dev) {
+    static bool init_first_run = true;
+    if (init_first_run) {
```

Comment:
This seems to be to avoid issues w/ multiple inits because you have both one shot keys and one shot layers as different behavior driver instances. Is there a reason they need to "share" the sticky key state between both?

Reply:
* this is to circument issues in `sticky_key_keycode_state_changed_listener`, where there is no access to the behaviors' data. (At least, I haven't found a way).
* Without the global `active_sticky_keys`, the 'data' struct needs to be pased to many methods.
* It also saves a little memory by having a single `active_sticky_keys` array that is shared by all instances. (it's btter to have a single big pool than multiple small pools)